test/known_problems/should_fail/poly_lists_map_should_fail.erl

```diff
@@ -0,0 +1,19 @@
+-module(poly_lists_map_should_fail).
+
+-gradualizer([solve_constraints]).
+
+-export([i/1,
+         j/1]).
+
+-spec i([binary() | integer()]) -> [integer()].
+i(L) ->
+    lists:map(fun
+                  (I) when is_integer(I) -> I * 2;
+                  (B) when is_list(B) -> list_to_integer(B)
+              end, L).
```

Comment:
Since these examples are slightly complex, it'd be interesting to see what errors message we expect here. Can you add it in a comment (or do you think it I should run it myself to find out)?

Reply:
Yeah, this one is a bit far fetched as it assumes smart enough inference of the inline fun's type. I imagine the error could look similar to the ones we currently report on this branch:

```
Lower bound binary() | integer() of type variable T_123 on line 456 is not a subtype of integer() | list()
```

We could also try to be smarter and infer the input type of the inline fun to be `integer() & list()`, but that's even more distant future.

However, more explicit examples would not rely on the inference of the inline fun's type at all, while still capture the problem. Let's take these which rely on just on top-level functions with specs:

```erlang
-spec i([binary() | integer() | string()]) -> [integer()].
i(L) ->
    lists:map(fun takes_a_union/1, L).

-spec takes_a_union(binary() | integer()) -> integer().
takes_a_union(B) when is_binary(B) -> binary_to_integer(B);
takes_a_union(I) when is_integer(I) -> I.
```

`i/1` should fail type checking, because the union it accepts is wider than `takes_a_union/1` allowed input type.

```erlang
-spec j([binary() | integer() | string()]) -> list().
j(L) ->
    lists:map(fun takes_an_intersection/1, L).

-spec takes_an_intersection(binary()) -> binary();
                           (integer()) -> integer().
takes_an_intersection(B) when is_binary(B) -> B;
takes_an_intersection(I) when is_integer(I) -> I.
```

`j/1` is similar, but it calls a function which takes an intersection type instead of a union.

I'll add these two simpler examples to tests, too.

Reply:
Great, just add that in a code comment then, something like

```Erlang
%% We expect an error like "Lower bound binary() | integer()
%% of type variable T_123 on line 456 is not a subtype of
%% integer() | list()"
-spec i...
```